test/e2e/apimachinery/crd_publish_openapi.go

```diff
@@ -833,3 +903,220 @@ properties:
+			portForwardRegexp, err := regexp.Compile(portForwardExpr)
+			if err != nil {
+				return false, fmt.Errorf("failed to compile a regexp for finding a local listening port, err: %v, expression: %s", err, portForwardExpr)
+			}
+			match := portForwardRegexp.FindStringSubmatch(portForwardOutput)
+			if len(match) != 3 {
+				framework.Logf("failed to parse kubectl port-forward output: %q, with %q reg exp, expected to find exactly 3 matches, found: %v", portForwardOutput, portForwardExpr, match)
+				return false, nil
+			}
+
+			localListenPort, err := strconv.Atoi(match[2])
+			if err != nil {
+				return false, fmt.Errorf("error converting %s to an int: %v", match[2], err)
+			}
+			localPorts[index] = localListenPort
+		}
+		if len(localPorts) != len(remotePorts) {
+			framework.Logf("not all ports have been forwarded, found %d forwarded ports but want %d", len(localPorts), len(remotePorts))
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		framework.Failf("Timeout waiting/parsing kubectl port-forward command's output: %v", err)
+	}
+
+	return &PortForwardCommand{
+		cmd:        cmd,
+		localPorts: localPorts,
+	}
+}
+
+// apiServersProxyPod creates a pod that once run proxies connections to the given apiServers.
+// It also returns a remote ports list for convenience that will be used by kubectl port-forward
+func apiServersProxyPod(apiServers []string) (*v1.Pod, []int, error) {
+	script := `
+       apk add socat
+       socat TCP-LISTEN:${LOCAL_LISTEN_PORT},fork TCP:${SERVER_IP}:${SERVER_PORT}
+`
+
+	proxyListenPorts := []int{}
+	proxyListenPort := 8443
+	containers := []v1.Container{}
+	for index, apiServer := range apiServers {
+		ipPort := strings.Split(apiServer, ":")
+		if len(ipPort) != 2 {
+			return nil, nil, fmt.Errorf("incorrect apiServer =%s, expected to find an IP address and port in the form of IP:PORT", apiServer)
+		}
+
+		r := strings.NewReplacer(
+			"${LOCAL_LISTEN_PORT}", fmt.Sprintf("%d", proxyListenPort),
+			"${SERVER_IP}", ipPort[0],
+			"${SERVER_PORT}", ipPort[1],
+		)
+		modifiedScript := r.Replace(script)
+
+		containers = append(containers, v1.Container{
+			Name: fmt.Sprintf("server-%d", index),
+			Ports: []v1.ContainerPort{
+				{ContainerPort: int32(proxyListenPort)},
+			},
+			Image:   "alpine", // TODO: build an image with socat and then use  imageutils.GetE2EImage(imageutils.BusyBox),
+			Command: []string{"/bin/sh", "-c"},
+			Args:    []string{modifiedScript},
+		})
+		proxyListenPorts = append(proxyListenPorts, proxyListenPort)
+		proxyListenPort++
+	}
+
+	name := "api-servers-proxy"
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name + "-" + string(uuid.NewUUID()),
+		},
+		Spec: v1.PodSpec{
+			Containers: containers,
+		},
+	}, proxyListenPorts, nil
+}
+
+func getAllAPIServersEndpoint(c k8sclientset.Interface) ([]string, error) {
+	eps, err := c.CoreV1().Endpoints(metav1.NamespaceDefault).Get(context.TODO(), "kubernetes", metav1.GetOptions{})
```

Comment:
It seems that endpoints for kubernetes service are populated by [EndpointReconciler](https://github.com/kubernetes/kubernetes/blob/master/pkg/master/reconcilers/reconcilers.go#L38). From what I understand it should be okay as the endpoints are filled by kube API servers unless the endpoint reconciler was turned off. Could you please double-check that?

Reply:
@liggitt have you had time to verify that?

Reply:
Running without the endpoint reconciler is a supported option (`--endpoint-reconciler-type=none`), as is using a node-port for the kubernetes master service  (`--kubernetes-service-node-port=...`). Switching a conformance test to go directly to endpoints instead of through the kubernetes service doesn't seem proper.

Reply:
I would expect there to be a shim in our test framework that correctly waits for HA configurations (an HA configuration should be able to be shown to be conformant) by getting api endpoints appropriately depending on what the config is.  I.e. `h := framework.NewAPIEndpointHelper(client); h.WaitForAll(func(<gets a client config>) (bool, error) { do a test with the client to see if x is true })` that handles that.

ukasz do you think you would be willing to articulate a version of the helper that works in each case jordan mentioned and hides this behavior when it's not possible (if no endpoints are set, the test may also need a way to clarify that)?

Everyone is instead doing this poorly, incorrectly, and inconsistently in the existing code - having a correct way to do at least do this for the apiserver is good. I don't think it's acceptable for us to punt on HA apiservers which is the recommended prod configuration of kube to prove that they are conformant in a reasonable way (and medium length sleeps may not be sufficient).  We definitely paper over this in other components (stuff running in deployment sized two) - so it would be nice if the underlying code could be adapted to something like the OpenShift ingress tests, or the upgrade tests that verify the data.  We still have a need for probalistic checking (i.e. if we're trying to verify a SLB directly)

Reply:
> ukasz do you think you would be willing to articulate a version of the helper that works in each case jordan mentioned and hides this behavior when it's not possible (if no endpoints are set, the test may also need a way to clarify that)?

@smarterclayton sure, I can start with a version that understands endpoints. Many thanks.

Reply:
https://github.com/kubernetes/kubernetes/pull/100585